pkg/bmc/bmc.go

```diff
@@ -0,0 +1,349 @@
+
+		return fmt.Errorf("redfish connection error: %w", err)
+	}
+
+	defer func() {
+		redfishClient.Logout()
+		cancel()
+	}()
+
+	sboot, err := redfishGetSystemSecureBoot(redfishClient, bmc.systemIndex)
+	if err != nil {
+		glog.V(100).Infof("Failed to get redfish system's secure boot: %v", err)
+
+		return fmt.Errorf("failed to get secure boot: %w", err)
+	}
+
+	if sboot.SecureBootEnable {
+		glog.V(100).Infof("Failed to enable secure boot: it is already enabled")
+
+		return fmt.Errorf("secure boot is already enabled")
+	}
+
+	sboot.SecureBootEnable = true
+
+	err = sboot.Update()
+	if err != nil {
+		glog.V(100).Infof("Failed to enable secure boot: %v", err)
+
+		return fmt.Errorf("failed to enable secure boot: %w", err)
+	}
+
+	return sboot.Update()
+}
+
+// SecureBootDisable disables the SecureBoot feature using the BMC's RedFish API endpoint.
+func (bmc *BMC) SecureBootDisable() error {
+	glog.V(100).Infof("Disabling secure boot from bmc's redfish endpoint")
+
+	redfishClient, cancel, err := redfishConnect(
+		bmc.host,
+		bmc.redfishUser.Name, bmc.redfishUser.Password,
+		bmc.timeOuts.Redfish)
+	if err != nil {
+		glog.V(100).Infof("Redfish connection error: %v", err)
+
+		return fmt.Errorf("redfish connection error: %w", err)
+	}
+
+	defer func() {
+		redfishClient.Logout()
+		cancel()
+	}()
+
+	sboot, err := redfishGetSystemSecureBoot(redfishClient, bmc.systemIndex)
+	if err != nil {
+		glog.V(100).Infof("Failed to get redfish system's secure boot: %v", err)
+
+		return fmt.Errorf("failed to get secure boot: %w", err)
+	}
+
+	if !sboot.SecureBootEnable {
+		glog.V(100).Infof("Failed to disable secure boot: it is already disabled")
+
+		return fmt.Errorf("secure boot is already disabled")
+	}
+
+	sboot.SecureBootEnable = true
+
+	err = sboot.Update()
+	if err != nil {
+		glog.V(100).Infof("Failed to disable secure boot: %v", err)
+
+		return fmt.Errorf("failed to disable secure boot: %w", err)
+	}
+
+	return sboot.Update()
+}
+
+// SystemForceReset performs a (non-graceful) forced system reset using redfish API.
+func (bmc *BMC) SystemForceReset() error {
+	glog.V(100).Infof("Forcing system reset from bmc's redfish endpoint")
+
+	redfishClient, cancel, err := redfishConnect(
+		bmc.host,
+		bmc.redfishUser.Name, bmc.redfishUser.Password,
+		bmc.timeOuts.Redfish)
+	if err != nil {
+		glog.V(100).Infof("Redfish connection error: %v", err)
+
+		return fmt.Errorf("redfish connection error: %w", err)
+	}
+
+	defer func() {
+		redfishClient.Logout()
+		cancel()
+	}()
+
+	system, err := redfishGetSystem(redfishClient, bmc.systemIndex)
+	if err != nil {
+		glog.V(100).Infof("Failed to get redfish system: %v", err)
+
+		return fmt.Errorf("failed to get redfish system: %w", err)
+	}
+
+	return system.Reset(redfish.ForceRestartResetType)
+}
+
+func redfishConnect(host, user, password string, sessionTimeout time.Duration) (
+	*gofish.APIClient, context.CancelFunc, error) {
+	gofishConfig := gofish.ClientConfig{
```

Comment:
Since we have to log out after every operation to prevent leaking sessions, would it be better to use basic auth on every request and avoid starting sessions?

Reply:
As per my readings on redfish [api](https://www.dmtf.org/sites/default/files/standards/documents/DSP0266_1.20.0.pdf)/[docs](https://www.dmtf.org/sites/default/files/Redfish_School-Sessions.pdf), basic auth is not recommended unless you're going to do single operations once in a while.

Some of the BMC's methods I've implemented do more than one operation in a very short period. For instance, enabling the Secure Boot in a system might require 3 redfish requests/operations: one to get the system, another to get that sytem's secure boot, and a third one to update the secure boot's status. I know that because I've read some of the gofish code I'm using... but some of its methods might change in future gofish releases and we might not notice that they're internally using/calling even more redifsh operations.

Perhaps, if you're really interested in having BasicAuth as an option, we can expand the BMC's New() method with extra flags using the options pattern (`New(..., DefaultTimeOuts, WithBasicAuth()`), to set that basic auth as the preferred method, in case you know that the methods you're going to use are single-operation ones.

Reply:
Thanks for the explanation! I agree that if there are multiple underlying API calls within a session then it'll be useful to have a session for those. I don't have any use cases myself where the extra latency for creating/deleting sessions would be a concern so no need to expand the `New()` method.